cpp/src/binning/bin.cu

```diff
@@ -0,0 +1,265 @@
+  LeftComparator m_left_comp{};    // Comparator used for left edges.
+  RightComparator m_right_comp{};  // Comparator used for right edges.
+};
+
+// Functor to identify rows that should be filtered out based on the sentinel set by
+// bin_finder::operator().
+struct filter_null_sentinel {
+  __device__ bool operator()(size_type i) { return i != NULL_VALUE; }
+};
+
+// Bin the input by the edges in left_edges and right_edges.
+template <typename T, typename LeftComparator, typename RightComparator>
+std::unique_ptr<column> bin(column_view const& input,
+                            column_view const& left_edges,
+                            column_view const& right_edges,
+                            null_order edge_null_precedence,
+                            rmm::mr::device_memory_resource* mr,
+                            rmm::cuda_stream_view stream)
+{
+  auto output = make_numeric_column(
+    data_type(type_to_id<size_type>()), input.size(), mask_state::UNALLOCATED, stream, mr);
+  auto output_mutable_view = output->mutable_view();
+
+  // These device column views are necessary for creating iterators that work
+  // for columns of compound types. The column_view iterators fail for compound
+  // types because they return raw pointers to the start of the data.
+  auto input_device_view       = column_device_view::create(input);
+  auto left_edges_device_view  = column_device_view::create(left_edges);
+  auto right_edges_device_view = column_device_view::create(right_edges);
+
+  // Compute the maximum shift required for either edge, then shift all the iterators appropriately.
+  size_type null_shift = max(left_edges.null_count(), right_edges.null_count());
+  auto left_begin      = left_edges_device_view->begin<T>();
+  auto left_end        = left_edges_device_view->end<T>();
+  auto right_begin     = right_edges_device_view->begin<T>();
+
+  if (edge_null_precedence == null_order::BEFORE) {
+    left_begin  = thrust::next(left_begin, null_shift);
+    right_begin = thrust::next(right_begin, null_shift);
+  } else {
+    left_end = thrust::prev(left_end, null_shift);
+  }
+
+  // If all the nulls are at the beginning, the indices found by lower_bound
+  // will be off by null_shift, but if they're at the end the indices will
+  // already be correct.
+  size_type index_shift = (edge_null_precedence == null_order::BEFORE) ? null_shift : 0;
+
+  if (input.has_nulls()) {
+    thrust::transform(
+      rmm::exec_policy(stream),
+      input_device_view->pair_begin<T, true>(),
+      input_device_view->pair_end<T, true>(),
+      output_mutable_view.begin<size_type>(),
+      bin_finder<T, decltype(left_edges_device_view->begin<T>()), LeftComparator, RightComparator>(
+        left_begin, left_end, right_begin, index_shift));
+  } else {
+    thrust::transform(
+      rmm::exec_policy(stream),
+      input_device_view->pair_begin<T, false>(),
+      input_device_view->pair_end<T, false>(),
```

Comment:
```suggestion
      input_device_view->begin<T>(),
      input_device_view->end<T>(),
```
Technically if there are no nulls, you can just use the non-pair iterators.
This is just a suggestion. The symmetry is nice. 
The performance is probably about the same for this.

Reply:
Yes, I considered this. In that case I could overload `bin_finder<...>::operator()` to accept either a pair or a single value, and have the pair version call the value version if it's not null. I'll leave this open for other reviewers to comment as well in case they think it would be worthwhile, perhaps @jrhemstad has an opinion.

Reply:
I implemented this and it works fine. The downside is that in the case where there are nulls it introduces an extra function call (from `operator(thrust::pair<T, bool>)` to `operator(T)`), so unless nvcc can reliably optimize out that out by inlining the function this change would slightly slow the null path while speeding up the non-null path. I can benchmark if there's interest. I don't know if it's significant enough to worry about.